SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,436 @@
+      return other.isDistinct(from: self)
+    }
+
+    return false
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+  
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+  
+  func isDistinct(from other: AccessPath) -> Bool {
+    if base.isDistinct(from: other.base) { return true }
+    return false
+  }
+}
+
+/// An `AccessStoragePath` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessStoragePath {
+  let storage: Value
+  
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  let path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessStoragePath?, base: AccessBase)
+
+enum EnclosingScope {
+  case scope(BeginAccessInst)
+  case base(AccessBase)
+}
+
+
+func canBeOperandOfIndexAddr(_ value: Value) -> Bool {
+  switch value {
+  case is IndexAddrInst, is RefTailAddrInst:
+    return true
+  default:
+    return false
+  }
+}
+
+fileprivate struct PointerIdentificationUseDefWalkerInternal : ValueUseDefWalker {
+  var addrType: Type!
+  var result: Value?
+
+  mutating func rootDef(value: Value, path: SmallProjectionPath) -> WalkResult {
+    self.result = nil
+    return .abortWalk
+  }
+
+  mutating func walkUp(value: Value, path: SmallProjectionPath) -> WalkResult {
+    switch value {
+    case is FunctionArgument:
+      self.result = value
+      return .continueWalk
+    case let atp as AddressToPointerInst:
+      if let result = result, atp.operand != result {
+        self.result = nil
+        return .abortWalk
+      }
+      // TODO: instead of checking type equality perhaps check
+      //       layout compatibility, if possible?
+      if addrType == atp.operand.type, path.isEmpty {
+        self.result = atp.operand
+        return .continueWalk
+      }
+    case is BlockArgument, is MarkDependenceInst, is CopyValueInst,
+         is StructExtractInst, is TupleExtractInst, is StructInst, is TupleInst:
+      return walkUpDefault(value: value, path: path)
+    default:
+      break
+    }
+
+    self.result = nil
+    return .abortWalk
+  }
+
+  var walkUpCache = WalkerCache<Path>()
+}
+
+struct PointerIdentificationWalker {
+  private var walker = PointerIdentificationUseDefWalkerInternal()
+
+  mutating func compute(_ atp: PointerToAddressInst) -> Value? {
+    walker.addrType = atp.type
+    walker.result = nil
+    _ = walker.walkUp(value: atp.operand, path: SmallProjectionPath())
+    return walker.result
+  }
+}
+
+fileprivate struct AccessPathWalkerInternal : AddressUseDefWalker {  
+  var result: AccessPath? = nil
+  private var _beginAccess: BeginAccessInst? = nil
```

Comment:
Why is this underscored? Also, I think `foundBeginAccess` would be a better name.